### ISSUE TITLE
Text within input with box-sizing: border-box is not positioned correctly

### ISSUE THREAD

**Issue by blacksvk** [servo/servo#11528]
![itext](https://cloud.githubusercontent.com/assets/10187350/15687637/1291db28-2776-11e6-88e1-8711a815787c.png)
```
<!doctype html>
<html><body>
	<input placeholder="Enter your mail" type="text" style="padding:0.5em; height: 2.2em; box-sizing: border-box;">
</body></html>
```

**Comment by jdm** [comment]
It's not yet clear to me whether this is another instance of our form-related CSS rules not overriding enough properties, or whether this is a bug in our layout implementation.

**Comment by blacksvk** [comment]
Ok, it seems like there are 2 issues:

1) text have different size, that's why box is smaller. When font-size is specified, box have correct size (compared to Firefox and another browsers.)

2) text within box is not positioned correctly

![fz](https://cloud.githubusercontent.com/assets/10187350/15688255/60695972-2779-11e6-964f-c7bd47e0a998.png)

I will rename this issue to original name - that's describing 2), which seems to be core of this problem.
Should I create separate issue for 1)?

**Comment by jdm** [comment]
Please do!

**Comment by mskrzypkows** [comment]
It's strange, in my FF (46.0.1v) on Ubuntu, text is displayed as in Servo. Is text position wrong here? Maybe it depends on used font.
![ff_ubuntu](https://cloud.githubusercontent.com/assets/10705549/15705894/63fd5048-27f1-11e6-8002-0f4ad2bd30f3.png)

**Comment by blacksvk** [comment]
That's strange. I've tested that on multiple browsers in Windows (including FF 46.0.1) and they render that like on the first picture.
![win](https://cloud.githubusercontent.com/assets/10187350/15706397/a7c8058c-27f3-11e6-9dbf-8ee6537ed7be.png)

I don't think it's font-related, because it's rendered differently even when you specify font-family and font-size.

@mskrzypkows Can you try how FF on Ubuntu render something like this?
`<input placeholder="Enter your mail" type="text" style="padding:1em; height: 10em; box-sizing: border-box; width: 20em; font-size: 1em; font-family: Arial;">`
On Windows it looks like this:
![t2](https://cloud.githubusercontent.com/assets/10187350/15706855/02501ab0-27f6-11e6-869c-9810fb8c87b3.png)

**Comment by mskrzypkows** [comment]
You're right, it seem that text position is wrong. I suppose that in the first example, text is cut because of native Ubuntu control.
![ff_ubuntu](https://cloud.githubusercontent.com/assets/10705549/15711722/99357448-280f-11e6-9836-803f99840c54.png)

**Comment by mskrzypkows** [comment]
@jdm I can investigate the issue, could you give me some hint, in which area should I look for rendering label of input?

**Comment by jdm** [comment]
The actual text content is derived from `ServoThreadSafeLayoutNode::text_content()` in `components/layout/wrapper.rs`. That's invoked from `FlowConstructor::create_fragments_for_node_text_content` in `components/layout/construct.rs`. There are rendering rules for text inputs in `resources/servo.css` that will apply here as well.

**Comment by jdm** [comment]
Also `resources/quirks-mode.css` has a rule that may apply depending on the testcase; I forget when quirks mode is applied.

**Comment by mskrzypkows** [comment]
The style: https://github.com/servo/servo/blob/master/resources/servo.css#L46 doesn't work. I've checked in FF, there is text always centred vertical in input. Maybe we need to add some special centering for inputs?

**Comment by stshine** [comment]
@highfive: assign me

**Comment by nox** [comment]
Looks better now but still no perfect:

<img width="1136" alt="capture d ecran 2017-10-01 a 18 15 50" src="https://user-images.githubusercontent.com/123095/31056614-96f3c622-a6d4-11e7-804d-4193dfb8849d.png">

**Comment by lumiscosity** [comment]
Fixed completely, now. Tested on Arch Linux x86_64, servo built against commit id 9d29017c0dc6eec4ca7cd868cc13056e0bc0c64a, compared against Firefox 141.0, Epiphany 48.5 and Chromium 138.0.7204.168 (ordered left to right).

<img width="1121" height="737" alt="Image" src="https://github.com/user-attachments/assets/3e40dee9-b5fa-4a99-80e3-f2a28f75e631" />

**Comment by jdm** [comment]
Yay!
